File: leetcode/0659-split-array-into-consecutive-subsequences/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean isPossible(int[] nums) {
        Map<Integer, Integer> count = new HashMap<>();
        for (int x : nums) count.merge(x, 1, Integer::sum);
        Map<Integer, Integer> tails = new HashMap<>();

        for (int x : nums) {
            if (count.getOrDefault(x, 0) == 0) continue;
            count.merge(x, -1, Integer::sum);

            if (tails.getOrDefault(x - 1, 0) > 0) {
                tails.merge(x - 1, -1, Integer::sum);
                tails.merge(x, 1, Integer::sum);
            } else if (count.getOrDefault(x + 1, 0) > 0 && count.getOrDefault(x + 2, 0) > 0) {
                count.merge(x + 1, -1, Integer::sum);
                count.merge(x + 2, -1, Integer::sum);
                tails.merge(x + 2, 1, Integer::sum);
            } else {
                return false;
            }
        }
        return true;
    }
}

File: leetcode/0659-split-array-into-consecutive-subsequences/solution.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool isPossible(vector<int>& nums) {
        unordered_map<int, int> count, tails;
        for (int x : nums) count[x]++;

        for (int x : nums) {
            if (count[x] == 0) continue;
            count[x]--;

            if (tails[x - 1] > 0) {
                tails[x - 1]--;
                tails[x]++;
            } else if (count[x + 1] > 0 && count[x + 2] > 0) {
                count[x + 1]--;
                count[x + 2]--;
                tails[x + 2]++;
            } else {
                return false;
            }
        }
        return true;
    }
};
